services/database/scripts/ContractVerifier.mjs

```diff
@@ -0,0 +1,342 @@
+    });
+    this.batchCount++;
+    this.offset += result.rows.length;
+
+    return result.rows;
+  }
+
+  async markContractAsVerified(chainId, address) {
+    const query = `
+      UPDATE missing_transaction_hash 
+      SET reverified = true, updated_at = NOW()
+      WHERE chain_id = $1 AND address = $2
+    `;
+    await this.pool.query(query, [chainId, address]);
+  }
+
+  async verifyContract(contract) {
+    const files = contract.sources.reduce((acc, source) => {
+      acc[source.path] = source.content;
+      return acc;
+    }, {});
+    files["metadata.json"] = contract.sourcify_metadata;
+
+    const body = {
+      chainId: contract.chain_id.toString(),
+      address: `0x${contract.address}`,
+      metadata: JSON.parse(contract.sourcify_metadata),
+      files,
+    };
+
+    return fetch(`${this.options.server}/verify`, {
+      method: "POST",
+      headers: { "Content-Type": "application/json" },
+      body: JSON.stringify(body),
+    });
+  }
+
+  async verifyAndMarkContract() {
+    const contract = this.currentContractsBatch.pop();
+    this.activeVerifications++;
+    const verifyStartTime = Date.now();
+
+    try {
+      logger.debug("⌛ Verifying contract", {
+        chainId: contract.chain_id,
+        address: `0x${contract.address}`,
+        activeVerifications: this.activeVerifications,
+        inVerificationCurrentBatch: `${this.options.batchSize - this.currentContractsBatch.length}/${this.options.batchSize}`,
+      });
+
+      const response = await this.verifyContract(contract);
+      await this.handleVerificationResponse(
+        response,
+        contract,
+        verifyStartTime,
+      );
+    } catch (err) {
+      logger.error("❌ Verification failed", {
+        chainId: contract.chain_id,
+        address: `0x${contract.address}`,
+        error: err.message,
+        durationSeconds: ((Date.now() - verifyStartTime) / 1000).toFixed(2),
+      });
+    } finally {
+      this.activeVerifications--;
+    }
+  }
+
+  async handleVerificationResponse(response, contract, verifyStartTime) {
+    if (response.ok) {
+      await this.handleSuccessfulVerification(contract, verifyStartTime);
+    } else if (response.status === 409) {
+      await this.handleAlreadyVerified(contract);
+    } else {
+      await this.handleFailedVerification(response, contract, verifyStartTime);
+    }
+  }
+
+  async handleSuccessfulVerification(contract, verifyStartTime) {
+    logger.debug("✅ Verified contract", {
+      chainId: contract.chain_id,
+      address: `0x${contract.address}`,
+      durationSeconds: ((Date.now() - verifyStartTime) / 1000).toFixed(2),
+      rateSinceStart: (
+        this.verifiedCount /
+        ((Date.now() - this.startTime) / 1000)
+      ).toFixed(2),
+    });
+    await this.markContractAsVerified(
+      contract.chain_id,
+      Buffer.from(contract.address, "hex"),
+    );
+    this.verifiedCount++;
+  }
+
+  async handleAlreadyVerified(contract) {
```

Comment:
Are we actually updating the creation transaction hash in the db if we get this error?

Reply:
No the scripts don't do manual edits to the DB, it has to be a full verification from scratch. 

The full workflow needs to be run multiple times. After running couple times until you get to reasonable leftover contracts, one needs to delete the table and create again (looking at the missing tx hashes in the DB). If a contract throwing this error was being verified, it will not show up again in the new table. Otherwise one needs to look at logs etc. why the contract is not receiving the creator tx hash.

Reply:
My questions was rather targeting the server. Does the server update the creation transaction hash in the database if the contract was already verified and the reverification doesn't lead to a better match?
If I read the `StorageService` code correctly, it throws before storing anything in that case.

Reply:
I think it implicitly does because it's actually a better match, from `creationMatch=null` to `creationMatch=partial/perfect`

Here 
https://github.com/ethereum/sourcify/blob/83cf76bda936530048401ef4787908ab0a4db961/services/server/src/server/services/utils/util.ts#L75-L99

and the check is used here:
https://github.com/ethereum/sourcify/blob/83cf76bda936530048401ef4787908ab0a4db961/services/server/src/server/services/StorageService.ts#L266-L292

Reply:
But I guess to your point, it does not save this info if it doesn't get a creation match. Question is, should it? Ideally yes but the amount of contracts is quite miniscule now. We only have around 600 contracts total left that can't be verified with a creation match

Reply:
Ah yes that's the point. It only doesn't if cannot get a creation match. I think it is fine then.